### ISSUE TITLE
api.internal dns record not updated on rolling-update

### ISSUE THREAD

**Issue by apenney** [kubernetes/kops#2634]
After my new masters were up the dns entry for api.internal was left to the 3 old masters, making all nodes break.  This was kops 1.6.0 and kubernetes 1.6.4.

**Comment by apenney** [comment]
I think this is related to #2633. Rumor is that the masters are responsible for updating this record, and mine are broken on start due to no flannel configuration.

**Comment by mtanski** [comment]
I was also affected by this issue.

The workaround is change the api.internal.* Route53 DNS entry to point to the internal ELB created by kops. This way the next rolling update / or ASG action (like recovering a failed master instance) does not leave you broken cluster where nodes cannot connect.

**Comment by chrislovecnm** [comment]
@apenny can I get more details?

1. Cluster setups
2. Exact steps taken

I assume this is kops 1.6.0

Would like to recreate ;)

**Comment by mtanski** [comment]
Steps:
1. Create a private cluster on AWS (no bastion needed)
2. Observer internal.api.${DOMAIN} Route53 A entry
3. Do a rolling update (force to make sure master also gets re-done)
4. Check the internal.api.${DOMAIN} entry still points to the old instance
5. Wait for nodes to stop being able to communicate with master

I'm not sure this happens first time (or every time) but eventually I'm able to make this happen.

Cluster info:
provider: aws 1 zone
type: private
CNI: weave
masters: 1
nodes: two IG with cluster-autoscaler for ASG

**Comment by chrislovecnm** [comment]
> Create a private cluster on AWS (no bastion needed)

k8s version?

> Observer internal.api.${DOMAIN} Route53 A entry

ELI5 please

> Do a rolling update (force to make sure master also gets re-done)

What version to what version of k8s?  Why force?  Any feature flags?  What is the command you are doing?

> type: private

topology or DNS?

**Comment by apenney** [comment]
I was going from 1.6.2 to 1.6.4.  When I did the rolling-update my master's came back up but did not update the DNS record for internal.api, leaving it pointed at the 3 ip's of my old masters.

Initially I thought the problem was missing 10-flannel.conf, but I think that would have generated if they could have reached internal.api.

Here's the cluster conf, with some domain hiding :)

```
apiVersion: kops/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: 2017-03-24T01:40:17Z
  name: prod.domain
spec:
  additionalPolicies:
    node: |
      [
        {
          "Action": ["sts:AssumeRole"],
          "Effect": "Allow",
          "Resource": ["*"]
        }
      ]
  api:
    loadBalancer:
      type: Public
  authorization:
    alwaysAllow: {}
  channel: alpha
  cloudProvider: aws
  configBase: s3://xxx/prod.domain
  dnsZone: xxx
  etcdClusters:
  - etcdMembers:
    - instanceGroup: master-us-east-1c
      name: c
    - instanceGroup: master-us-east-1d
      name: d
    - instanceGroup: master-us-east-1e
      name: e
    name: main
  - etcdMembers:
    - instanceGroup: master-us-east-1c
      name: c
    - instanceGroup: master-us-east-1d
      name: d
    - instanceGroup: master-us-east-1e
      name: e
    name: events
  kubeAPIServer:
    runtimeConfig:
      batch/v2alpha1: "true"
  kubernetesApiAccess:
  - 0.0.0.0/0
  kubernetesVersion: 1.6.2
  masterInternalName: api.internal.prod.domain
  masterPublicName: api.prod.domain
  networkCIDR: 10.10.0.0/16
  networkID: vpc-37917a50
  networking:
    flannel: {}
  nonMasqueradeCIDR: 100.64.0.0/10
  sshAccess:
  - 0.0.0.0/0
  subnets:
  - cidr: 10.10.64.0/19
    name: us-east-1c
    type: Private
    zone: us-east-1c
  - cidr: 10.10.96.0/19
    name: us-east-1d
    type: Private
    zone: us-east-1d
  - cidr: 10.10.128.0/19
    name: us-east-1e
    type: Private
    zone: us-east-1e
  - cidr: 10.10.160.0/22
    name: utility-us-east-1c
    type: Utility
    zone: us-east-1c
  - cidr: 10.10.164.0/22
    name: utility-us-east-1d
    type: Utility
    zone: us-east-1d
  - cidr: 10.10.168.0/22
    name: utility-us-east-1e
    type: Utility
    zone: us-east-1e
  topology:
    bastion:
      bastionPublicName: bastion.prod.domain
    dns:
      type: Public
    masters: private
    nodes: private
```

**Comment by justinsb** [comment]
Do you have the logs from the dns-controller pod?  Hopefully there's an error in there (or maybe it's not running, in which case we can figure out why that is!)

**Comment by apenney** [comment]
I was going to try and get that log for you but I can't even find a dns-controller container on the masters in this state:

```
CONTAINER ID        IMAGE                                                                                                                      COMMAND                  CREATED              STATUS              PORTS               NAMES
c7a9b4267109        gcr.io/google_containers/kube-controller-manager@sha256:a93d4c26d71de94861f78cf5ea62600e4952685d580e2774c630ea206b7c18ee   "/bin/sh -c '/usr/loc"   About a minute ago   Up About a minute                       k8s_kube-controller-manager_kube-controller-manager-ip-10-10-115-156.ec2.internal_kube-system_6332a0382cba0d3cb296642195686e33_1
472af271e1f0        gcr.io/google_containers/kube-scheduler@sha256:57661c79890b01ef2ff183ed4b467ca470efc4fb8d0517cd29abe49e72f6d904            "/bin/sh -c '/usr/loc"   About a minute ago   Up About a minute                       k8s_kube-scheduler_kube-scheduler-ip-10-10-115-156.ec2.internal_kube-system_bdcc53ecea66fdfbc80f90cb2040efe0_1
1e6a7cc19849        sysdig/agent@sha256:c515a0f8a598f761b1d95a99b2c39e41974102b816db9baf1456738bc395eb67                                       "/docker-entrypoint.s"   13 minutes ago       Up 13 minutes                           k8s_sysdig_honking-uakari-sysdig-w0ptv_default_d33e6c3d-4298-11e7-b973-0aa462458000_0
9ca8900b1513        gcr.io/google_containers/pause-amd64:3.0                                                                                   "/pause"                 13 minutes ago       Up 13 minutes                           k8s_POD_honking-uakari-sysdig-w0ptv_default_d33e6c3d-4298-11e7-b973-0aa462458000_0
dde60ea248c1        gcr.io/google_containers/kube-apiserver@sha256:6d5aa429c2b0806e4b6d1d179054d6deee46eec0aabe7bd7bd6abff97be36ae7            "/bin/sh -c '/usr/loc"   14 minutes ago       Up 14 minutes                           k8s_kube-apiserver_kube-apiserver-ip-10-10-115-156.ec2.internal_kube-system_cffca12d75806fc9ecda845ba11b6591_0
49385f2decaa        gcr.io/google_containers/etcd@sha256:19544a655157fb089b62d4dac02bbd095f82ca245dd5e31dd1684d175b109947                      "/bin/sh -c '/usr/loc"   14 minutes ago       Up 14 minutes                           k8s_etcd-container_etcd-server-ip-10-10-115-156.ec2.internal_kube-system_aca7550e484564346fb9515deba574e7_0
21676fb20a30        gcr.io/google_containers/kube-proxy@sha256:44cc08e7e8a2089eb8dfad6b692e9ece5994d6e6cff07fc9e9b1273cab0f6c6a                "/bin/sh -c 'echo -99"   14 minutes ago       Up 14 minutes                           k8s_kube-proxy_kube-proxy-ip-10-10-115-156.ec2.internal_kube-system_ed40c6c30d5531642200e59c0fd91fe4_0
2101a78b8104        gcr.io/google_containers/etcd@sha256:19544a655157fb089b62d4dac02bbd095f82ca245dd5e31dd1684d175b109947                      "/bin/sh -c '/usr/loc"   14 minutes ago       Up 14 minutes                           k8s_etcd-container_etcd-server-events-ip-10-10-115-156.ec2.internal_kube-system_1c994f9301c3cfabb3eb99c425eaecb9_0
f726eadbcc14        gcr.io/google_containers/pause-amd64:3.0                                                                                   "/pause"                 14 minutes ago       Up 14 minutes                           k8s_POD_kube-apiserver-ip-10-10-115-156.ec2.internal_kube-system_cffca12d75806fc9ecda845ba11b6591_0
d6c4ac51a56c        gcr.io/google_containers/pause-amd64:3.0                                                                                   "/pause"                 14 minutes ago       Up 14 minutes                           k8s_POD_etcd-server-events-ip-10-10-115-156.ec2.internal_kube-system_1c994f9301c3cfabb3eb99c425eaecb9_0
3c4b2c4ce2f2        gcr.io/google_containers/pause-amd64:3.0                                                                                   "/pause"                 14 minutes ago       Up 14 minutes                           k8s_POD_etcd-server-ip-10-10-115-156.ec2.internal_kube-system_aca7550e484564346fb9515deba574e7_0
98610e1350bc        gcr.io/google_containers/pause-amd64:3.0                                                                                   "/pause"                 14 minutes ago       Up 14 minutes                           k8s_POD_kube-proxy-ip-10-10-115-156.ec2.internal_kube-system_ed40c6c30d5531642200e59c0fd91fe4_0
6f7e4b30e213        gcr.io/google_containers/pause-amd64:3.0                                                                                   "/pause"                 14 minutes ago       Up 14 minutes                           k8s_POD_kube-scheduler-ip-10-10-115-156.ec2.internal_kube-system_bdcc53ecea66fdfbc80f90cb2040efe0_0
bce0e68c9630        gcr.io/google_containers/pause-amd64:3.0                                                                                   "/pause"                 14 minutes ago       Up 14 minutes                           k8s_POD_kube-controller-manager-ip-10-10-115-156.ec2.internal_kube-system_6332a0382cba0d3cb296642195686e33_0
ca786bda16cb        protokube:1.6.0                                                                                                            "/usr/bin/protokube -"   14 minutes ago       Up 14 minutes                           high_roentgen
```

**Comment by apenney** [comment]
Also to confirm, just fixing api.internal's dns wasn't enough for my masters to recover.  I had to manually build /etc/cni/net.d/10-flannel.conf on them all.

**Comment by apenney** [comment]
I think this may be related to the fact dns-controller cannot schedule on these broken clusters:

```
» kubectl describe po/dns-controller-309321819-fpm54 -n kube-system
Name:		dns-controller-309321819-fpm54
Namespace:	kube-system
Node:		/
Labels:		k8s-addon=dns-controller.addons.k8s.io
		k8s-app=dns-controller
		pod-template-hash=309321819
		version=v1.6.1
Annotations:	kubernetes.io/created-by={"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"ReplicaSet","namespace":"kube-system","name":"dns-controller-309321819","uid":"540e10c9-456a-11e7-9c02-0ad...
		scheduler.alpha.kubernetes.io/critical-pod=
		scheduler.alpha.kubernetes.io/tolerations=[{"key": "dedicated", "value": "master"}]
Status:		Pending
IP:
Controllers:	ReplicaSet/dns-controller-309321819
Containers:
  dns-controller:
    Image:	kope/dns-controller:1.6.1
    Port:
    Command:
      /usr/bin/dns-controller
      --dns=aws-route53
      --zone=*/Z1UD7P19NHNXZ2
      --zone=*/*
      -v=2
    Requests:
      cpu:		50m
      memory:		50Mi
    Environment:	<none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from dns-controller-token-0tb0f (ro)
Conditions:
  Type		Status
  PodScheduled 	False
Volumes:
  dns-controller-token-0tb0f:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	dns-controller-token-0tb0f
    Optional:	false
QoS Class:	Burstable
Node-Selectors:	node-role.kubernetes.io/master=
Tolerations:	node.alpha.kubernetes.io/notReady=:Exists:NoExecute for 300s
		node.alpha.kubernetes.io/unreachable=:Exists:NoExecute for 300s
Events:
  FirstSeen	LastSeen	Count	From			SubObjectPath	Type		Reason			Message
  ---------	--------	-----	----			-------------	--------	------			-------
  28s		13s		6	default-scheduler			Warning		FailedScheduling	No nodes are available that match all of the following predicates:: MatchNodeSelector (6), PodToleratesNodeTaints (3).
```

It seems to be missing:

```
Tolerations:    node-role.kubernetes.io/master=:NoSchedule
```

**Comment by apenney** [comment]
@mtanski https://github.com/kubernetes/kops/issues/2594 - seems to be this issue for me.

**Comment by chrislovecnm** [comment]
@apenny can we close this since the other issue is open?

**Comment by killianbrackey** [comment]
I also ran into this same issue with a forced rolling update (lost connectivity to masters eventually due to the route53 records not being updated).  Running with kops 1.6 and kubernetes 1.6.2 on kubenet.  Fixed by manually updating the route53 records

**Comment by mtanski** [comment]
@chrislovecnm We can close this issue if the scope of the linked issue is expanded. "dns controller rs fails to run pod after upgrade from 1.5 to 1.6.2" does not sufficiently capture the cases.

Just a rolling update of masters or auto-replaced master from ASG can trigger it.

**Comment by drequena** [comment]
Same problem over here!
Kops 1.8.1 - K8S 1.8.6 - kubectl 1.8.8
After "kops rolling-update cluster --yes" the process stops because the API Route53 record is not updated to the new IP.
Is there a work around? or any troubleshooting process?
Thanks.
